src/_P036_FrameOLED.ino

```diff
@@ -980,37 +1035,73 @@ byte display_scroll(int lscrollspeed, int lTaskTimer)
-    if ((PixLength > SizeSettings[OLEDIndex].Width) && (fScrollTime > 0.0))
-    {
-      // width of the line > display width -> scroll line
-      ScrollingLines.Line[j].Content = ScrollingPages.newString[j];
-      ScrollingLines.Line[j].Width = PixLength;
-      ScrollingLines.Line[j].CurrentLeft = SizeSettings[OLEDIndex].PixLeft;
-      ScrollingLines.Line[j].fPixSum = (float) SizeSettings[OLEDIndex].PixLeft;
-      ScrollingLines.Line[j].dPix = ((float)(PixLength-SizeSettings[OLEDIndex].Width))/fScrollTime; // pix change per scrolling line tick
-
-// log += F(" line: ");
-// log += j+1;
-// log += F(" width: ");
-// log += ScrollingLines.Line[j].Width;
-// log += F(" dPix: ");
-// log += ScrollingLines.Line[j].dPix;
-    }
-    else {
-      ScrollingLines.Line[j].Width = 0; // do not scroll line
+    word MaxPixWidthForPageScrolling = P36_MaxDisplayWidth;
+    if (bScrollLines) {
+      // Reduced scrolling width because line is displayed left or right aligned
+      MaxPixWidthForPageScrolling -= SizeSettings[OLEDIndex].PixLeft;
     }
+
+    for (byte j = 0; j < ScrollingPages.linesPerFrame; j++)
+    {
+      // default no line scrolling and strings are centered
+      ScrollingLines.Line[j].LastWidth = 0;
+      ScrollingLines.Line[j].Width = 0;
+
+      // get last and new line width
+      word LastPixLength = display->getStringWidth(ScrollingPages.oldString[j]);
+      word NewPixLength = display->getStringWidth(ScrollingPages.newString[j]);
+
+      if (bScrollLines) {
+        // settings for following line scrolling
+        if (LastPixLength > SizeSettings[OLEDIndex].Width)
+          ScrollingLines.Line[j].LastWidth = LastPixLength;   // while page scrolling this line is right aligned
+
+        if ((NewPixLength > SizeSettings[OLEDIndex].Width) && (fScrollTime > 0.0))
+        {
+          // width of the line > display width -> scroll line
+          ScrollingLines.Line[j].Content = ScrollingPages.newString[j];
+          ScrollingLines.Line[j].Width = NewPixLength;   // while page scrolling this line is left aligned
+          ScrollingLines.Line[j].CurrentLeft = SizeSettings[OLEDIndex].PixLeft;
+          ScrollingLines.Line[j].fPixSum = (float) SizeSettings[OLEDIndex].PixLeft;
+          ScrollingLines.Line[j].dPix = ((float)(NewPixLength-SizeSettings[OLEDIndex].Width))/fScrollTime; // pix change per scrolling line tick
+
+          // log += F(" line: ");
+          // log += j+1;
+          // log += F(" width: ");
+          // log += ScrollingLines.Line[j].Width;
+          // log += F(" dPix: ");
+          // log += ScrollingLines.Line[j].dPix;
+        }
+      }
+
+      // reduce line content for page scrolling to max width
+      while (NewPixLength > MaxPixWidthForPageScrolling) {
+        // shorten string on right side because line is displayed left aligned while scrolling
+        ScrollingPages.newString[j] = ScrollingPages.newString[j].substring(0, ScrollingPages.newString[j].length()-1);
```

Comment:
Can't this be made a bit more intelligent, like computing the offset needed and apply that?
Calling `substring` is quite an expensive operation and computing `length()` each time will probably yield very predictable results during the while loop.
I do understand there is still the need to let the library compute the pix length, but I guess you can also use a binary search like approach instead of subtracting 1 character.
For example, if a string of X characters needs Y pixels, then each character has on average Y/X pixels.
Just take the number of pixels it is too wide and divide that by your average width and round it down to get a good estimate on number of characters to throw away. If it is still too high, remove one or else add one.
Maybe you should also keep the old string and new string and just add the range to display of their string.
This way you have a good starting point for the new view and a lot less heap allocations needed.

Reply:
Done!